### PR TITLE
Preserve the formatting of mailmerge fields on merging

### PR DESCRIPTION
**Pull request by schtibe**
@sliverc

### CODE REVIEW

File: tests/test_winword2010.py [hunk 1]
```diff
@@ -24,34 +24,157 @@ def test(self):
...
-            'w:rsidRDefault="00886208" /><w:p w:rsidR="00886208" w:rsidRDefault="00886208"><w:r><w:t>I hope this '
-            'document from WinWord 2010 finds you well.</w:t></w:r></w:p><w:p w:rsidR="00886208" '
-            'w:rsidRDefault="00886208" /><w:p w:rsidR="00886208" w:rsidRDefault="00886208"><w:r><w:t>Kind regards,'
-            '</w:t></w:r></w:p><w:p w:rsidR="00886208" w:rsidRDefault="00886208" '
-            'w:rsidRPr="00886208"><w:pPr><w:rPr><w:lang w:val="en-US" /></w:rPr></w:pPr><w:proofErr '
-            'w:type="spellStart" /><w:r><w:t>docx-mailmerge</w:t></w:r><w:proofErr w:type="spellEnd" /><w:r><w:t>'
-            '.</w:t></w:r></w:p><w:sectPr w:rsidR="00886208" w:rsidRPr="00886208"><w:pgSz w:h="16838" w:w="11906" '
-            '/><w:pgMar w:bottom="1417" w:footer="708" w:gutter="0" w:header="708" w:left="1417" w:right="1417" '
-            'w:top="1417" /><w:cols w:space="708" /><w:docGrid w:linePitch="360" /></w:sectPr></w:body></w:document>')
+            '<w:document xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" mc:Ignorable="w14 wp14">'  # noqa
```
Comment (schtibe) [inline tests/test_winword2010.py:27]: I thought I'd leave it formatted, makes it more readable!

**Comment by sliverc** [comment]
The code currently expects that the instrText node is the last node and formatting tags are before.

Not sure whether this exists but if you have a construct where there are tags before and after instText current code will reorder the xml:

Example:

```
  <w:r w:rsidRPr="008B6CBB">
        <w:rPr>
          <w:b/>
        </w:rPr>
        <w:instrText xml:space="preserve"> MERGEFIELD Anrede </w:instrText>
       <w:rPr>
          <w:b/>
        </w:rPr>
      </w:r>
```

Will be:

```
 <w:r w:rsidRPr="008B6CBB">
        <w:rPr>
          <w:b/>
        </w:rPr>
       <w:rPr>
          <w:b/>
        </w:rPr>
        <w:it> MERGEFIELD Anrede </w:t>
      </w:r>
```

So in such cases we adjust the xml and Microsoft Word might end up with an error.

So wouldn't it be better to keep the whole construct as is and then simply replacing the instText with t tag?

File: mailmerge.py [hunk 2]
```diff
@@ -57,13 +57,21 @@ def __init__(self, file):
...
                 for idx_begin, idx_end, instr in fields:
-                    m = r.match(instr)
+                    m = r.match(instr.text)
                     if m is None:
                         continue
                     parent[idx_begin] = Element('MergeField', name=m.group(1))
+
+                    # append the other tags in the w:r block too
+                    block = instr.getparent()
+                    block.remove(instr)
```
Comment (schtibe) [inline mailmerge.py]: @sliverc check it out: it gets the parent of <instr> which is the <r> block. Then it throws away the <instr> and appends the rest to mergefield. So basically it takes all siblings of <instr> and preserves them, not considering any order of the tags.

Comment (schtibe) [inline mailmerge.py]: @sliverc ok, I understood you wrong. You actually _want_ to preserve the exact order.

**Comment by schtibe** [comment]
@sliverc please review the code again **wait** tests don't run

File: mailmerge.py [hunk 3]
```diff
@@ -152,10 +160,19 @@ def merge(self, parts=None, **replacements):
...
             mf.tag = '{%(w)s}r' % NAMESPACES
-            mf.append(Element('{%(w)s}t' % NAMESPACES))
-            mf[0].text = text
+            mf.extend(children)
+
+            text_node = Element('{%(w)s}t' % NAMESPACES)
+            text_node.text = text
+
+            ph = mf.find('MergeText')
+            if ph is not None:
```
Comment (sliverc) [inline mailmerge.py:173]: Could we add a comment here? It is not quite obvious why there is a if else as the code which makes this distinction is in a different function.

**Comment by sliverc** [comment]
Besides above comment code looks good. Also tested it with some files and seems to work smoothly.

**Comment by schtibe** [comment]
@sliverc did you get a notification now?

**Comment by sliverc** [comment]
Looks good and yes I got the notification.

**Pull request merged**